extensions/cl_intel_unified_shared_memory.asciidoc

```diff
@@ -0,0 +1,1264 @@
+
+===== Freeing Allocations
+
+The functions
+
+[source]
+----
+cl_int  clMemFreeINTEL(
+            cl_context context,
+            void* ptr);
+
+cl_int  clMemBlockingFreeINTEL(
+            cl_context context,
+            void* ptr);
+----
+
+free a Unified Shared Memory allocation.
+
+_context_ is a valid OpenCL context used to free the Unified Shared Memory allocation.
+
+_ptr_ is the Unified Shared Memory allocation to free.
+It must be a value returned by *clHostMemAllocINTEL*, *clDeviceMemAllocINTEL*, or *clSharedMemAllocINTEL*, or a `NULL` pointer.
+If _ptr_ is `NULL` then no action occurs.
+
+Note that *clMemFreeINTEL* may not wait for previously enqueued commands that may be using _ptr_ to finish before freeing _ptr_.
+It is the responsibility of the application to make sure enqueued commands that use _ptr_ are complete before freeing _ptr_.
+Applications should take particular care freeing memory allocations with kernels that may access memory indirectly, since a kernel with indirect memory access counts as using all memory allocations of the specified type or types.
+To wait for previously enqueued commands to finish that may be using _ptr_ before freeing _ptr_, use the *clMemBlockingFreeINTEL* function instead.
+
+*clMemFreeINTEL* and *clMemBlockingFreeINTEL* will return `CL_SUCCESS` if the function executes successfully.
+Otherwise, they will return one of the following error values:
+
+* `CL_INVALID_CONTEXT` if _context_ is not a valid context.
+* `CL_INVALID_VALUE` if _ptr_ is not a value returned by *clHostMemAllocINTEL*, *clDeviceMemAllocINTEL*, *clSharedMemAllocINTEL*, or a `NULL` pointer.
+* `CL_OUT_OF_RESOURCES` if there is a failure to allocate resources required by the OpenCL implementation on the device.
+* `CL_OUT_OF_HOST_MEMORY` if there is a failure to allocate resources required by the OpenCL implementation on the host.
+
+===== Controlling Allocations
+
+The table below describes allocation properties that may be passed to control allocation behavior.
+
+[[cl_mem_properties_intel]]
+.List of Supported `cl_mem_properties_intel` Properties
+[width="100%",cols="1,1,2",options="header"]
+|====
+| Property
+| Property Type
+| Description
+
+| `CL_MEM_ALLOC_FLAGS_INTEL`
+  | cl_mem_alloc_flags_intel
+    | Flags specifying allocation and usage information.
+      This is a bitfield type that may be set to a combination of the following values:
+
+      `CL_MEM_ALLOC_WRITE_COMBINED_INTEL`:
+      Request write combined (WC) memory.
+      Write combined memory may improve performance in some cases, however write combined memory must be used with care since it may hurt performance in other cases or use different coherency protocols than non-write combined memory.
+
+      `CL_MEM_ALLOC_INITIAL_PLACEMENT_DEVICE_INTEL`:
+      Request the implementation to optimize for first access being done by the device.
+      This flag is valid only for *clSharedMemAllocINTEL*.
```

Comment:
This states that the flag is valid only for `clSharedMemAllocINTEL`, but I only see error wording for it's usage under `clHostMemAllocINTEL`. Should this wording be under `clSharedMemAllocINTEL` ?

> CL_INVALID_PROPERTY if both CL_MEM_ALLOC_INITIAL_PLACEMENT_DEVICE_INTEL and CL_MEM_ALLOC_INITIAL_PLACEMENT_HOST_INTEL flags are specified.

Reply:
Good catch, yes, this error condition should definitely be under `clSharedMemAllocINTEL` and not `clHostMemAllocINTEL`.  I've moved it.

I'm also following up to see if any error conditions also apply to `clHostMemAllocINTEL` and `clDeviceMemAllocINTEL` or if the "valid only for `clSharedMemAllocINTEL`" phrase needs to be weakened.

Thanks for the careful review!

Reply:
I've also added similar error conditions for `clHostMemAllocINTEL` and `clDeviceMemAllocINTEL` if either flag is specified:

> * `CL_INVALID_PROPERTY` if either the `CL_MEM_ALLOC_INITIAL_PLACEMENT_DEVICE_INTEL` or `CL_MEM_ALLOC_INITIAL_PLACEMENT_HOST_INTEL` flags are specified.

I don't think our implementation is checking for these conditions right now, but I think an error makes the most sense since the initial placement flags do not apply to host or device allocations.

Thanks again!

Reply:
Makes sense, thanks for following up about this